lib/StackAnalysis/IndirectBranchInfoPrinterPass.cpp

```diff
@@ -0,0 +1,82 @@
+/// \file IndirectBranchInfoPrinterPass.cpp
+/// \brief Serialize the results of the StackAnalysis on disk.
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/InstrTypes.h"
+#include "llvm/Support/CommandLine.h"
+
+#include "revng/StackAnalysis/IndirectBranchInfoPrinterPass.h"
+#include "revng/Support/IRHelpers.h"
+
+using namespace llvm;
+
+static cl::opt<std::string> IndirectBranchInfoPath("indirect-branch-info-"
+                                                   "summary",
+                                                   cl::value_desc("filename"),
+                                                   cl::desc("Write the results "
+                                                            "of SA2 on disk."));
+
+void IndirectBranchInfoPrinterPass::serialize(
+  raw_ostream &OS,
+  CallBase *Call,
+  SmallVectorImpl<GlobalVariable *> &ABIRegs) {
+  auto Text = MemoryBuffer::getFileAsStream(IndirectBranchInfoPath);
+  if (!Text.getError()) {
+    if (!Text->get()->getBuffer().startswith("name")) {
+      OS << "name,ra,fso,address";
+      for (const auto &Reg : ABIRegs)
+        OS << "," << Reg->getName();
+      OS << "\n";
+    }
+  }
+
+  OS << Call->getParent()->getParent()->getName();
+  for (unsigned I = 0; I < Call->getNumArgOperands(); ++I) {
+    if (isa<ConstantInt>(Call->getArgOperand(I))) {
+      OS << "," << cast<ConstantInt>(Call->getArgOperand(I))->getSExtValue();
+    } else if (isa<StructType>(Call->getArgOperand(I)->getType())) {
+      auto PC = MetaAddress::fromConstant(Call->getArgOperand(I));
+      OS << "," << PC.address();
+    } else {
+      OS << ",unknown";
+    }
+  }
+  OS << "\n";
+}
+
+PreservedAnalyses
+IndirectBranchInfoPrinterPass::run(Function &F, FunctionAnalysisManager &FAM) {
+  if (IndirectBranchInfoPath.empty())
+    return PreservedAnalyses::all();
+
+  // Get the result of the GCBI analysis
+  auto &M = *F.getParent();
+  const auto &MAMProxy = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F);
+  GCBI = MAMProxy.getCachedResult<GeneratedCodeBasicInfoAnalysis>(M);
+  if (!GCBI)
+    GCBI = &(FAM.getResult<GeneratedCodeBasicInfoAnalysis>(F));
+  revng_assert(GCBI != nullptr);
+
+  SmallVector<GlobalVariable *, 16> ABIRegisters;
+  for (auto *CSV : GCBI->abiRegisters())
+    if (CSV && !(GCBI->isSPReg(CSV)))
+      ABIRegisters.emplace_back(CSV);
+
+  std::error_code EC;
+  raw_fd_ostream OS(IndirectBranchInfoPath,
+                    EC,
+                    sys::fs::OF_Append | sys::fs::OF_Text);
```

Comment:
Mmmh, I see why you're doing this. I think this should just be a module pass and you should drop `OF_Append` (and `OF_Text`?).

Reply:
I had thought about this, however I kept`IndirectBranchInfoPrinterPass` as a function pass, given that is used only in the SA2 pipeline, and the whole pipeline leverages a `FunctionPassManager`. Making it a module pass would require to instantiate a  `ModulePassManager`  + `ModuleAnalysisManager`. I preferred using `OF_Append` in favour of a simpler pipeline.

Reply:
Mh, the pass should have some "initialize" method. You could use that (or the constructor) to delete the file, so that the file only contains things relevant to this run. Otherwise we might accumulate stuff from different runs (not nice).

What about `OF_Text`?

Reply:
But the pass is run on each function, so the constructor / init is, no? Sounds not super good, but we probably need to switch a module pass.

Reply:
> But the pass is run on each function, so the constructor / init is, no?

Test it.

Reply:
Fixed, though I still need a static boolean variable as field of the class, since the constructor, as I suspected, is called at every run of the FPM.

Reply:
`static` fields are basically global variables, we don't really want them.

Now that I think about this, you might just use a `Logger`.
This seems pretty regular debug information we dump using loggers.

Reply:
Sounds good.